src/generate.ts

```diff
@@ -0,0 +1,385 @@
+    }
+  );
+
+  return output;
+}
+
+/*
+*********************
+* SIMPLE TYPE NODES *
+*********************
+
+* Simple types are either enums or primitives
+    * Example of a simple type as an enum:
+        <simpleType name="DeployProblemType">
+            <restriction base="string">
+                <enumeration value="Warning"/>
+                <enumeration value="Error"/>
+                <enumeration value="Info"/>
+            </restriction>
+        </simpleType>
+
+    * Examples of a simple type as a primitive, with and without additional restrictions, respectively:
+        <simpleType name="ID">
+            <restriction base="string">
+                <length value="18"/>
+                <pattern value="[a-zA-Z0-9]{18}"/>
+            </restriction>
+        </simpleType>
+        <simpleType name="PlatformSchemaContentType">
+            <restriction base="string"/>
+        </simpleType>
+
+* Note that restrictions like length and pattern are not taken into account while building the Typescript type.
+*/
+
+/*
+**********************
+* COMPLEX TYPE NODES *
+**********************
+
+* Complex type nodes contain either a sequence block with a number of element nodes representing attributes or a complex content block.
+    * Example of a complex type node with a sequence:
+        <xsd:complexType name="CompileAndTestRequest">
+            <xsd:sequence>
+                <xsd:element name="checkOnly" type="xsd:boolean"/>
+                <xsd:element name="classes" minOccurs="0" maxOccurs="unbounded" type="xsd:string"/>
+                <xsd:element name="deleteClasses" minOccurs="0" maxOccurs="unbounded" type="xsd:string"/>
+                <xsd:element name="deleteTriggers" minOccurs="0" maxOccurs="unbounded" type="xsd:string"/>
+                <xsd:element name="runTestsRequest" minOccurs="0" type="tns:RunTestsRequest"/>
+                <xsd:element name="triggers" minOccurs="0" maxOccurs="unbounded" type="xsd:string"/>
+            </xsd:sequence>
+        </xsd:complexType>
+
+    * Example of a complex type node with a complex content:
+        <xsd:complexType name="AIApplication">
+            <xsd:complexContent>
+                <xsd:extension base="tns:Metadata">
+                    <xsd:sequence>
+                        <xsd:element name="developerName" type="xsd:string"/>
+                        <xsd:element name="masterLabel" minOccurs="0" type="xsd:string"/>
+                        <xsd:element name="predictionDefinitions" minOccurs="0" maxOccurs="unbounded" type="tns:AIPredictionDefinition"/>
+                        <xsd:element name="status" type="tns:AIApplicationStatus"/>
+                        <xsd:element name="type" type="tns:AIApplicationType"/>
+                    </xsd:sequence>
+                </xsd:extension>
+            </xsd:complexContent>
+        </xsd:complexType>
+*/
+
+function treatComplexTypeNode(source: Array<SimpleTypeNode | ComplexTypeNode | ElementNode>): string {
```

Comment:
this is pretty long.  It feels like 2 composed functions

1. buildTypeMap(source)
2. typeMapToOutput

Reply:
and those *might* should be tested (this might be a good case for non-fs snapshot tests if the text is getting too long)

Reply:
all the comments (139-199) could be test cases 😄